lib/macroni/Conversion/Kernel/KernelRewriters.cpp

```diff
@@ -151,8 +176,144 @@ mlir::LogicalResult rewrite_rcu_dereference(macroni::MacroExpansion exp,
   }
 
   auto p_clone = rewriter.clone(*p);
-  rewriter.replaceOpWithNewOp<RCUDereference>(exp, exp.getType(0),
-                                              p_clone->getResult(0));
+  if (has_name(exp, "rcu_dereference")) {
+    rewriter.replaceOpWithNewOp<RCUDereference>(exp, exp.getType(0),
+                                                p_clone->getResult(0));
+  } else if (has_name(exp, "rcu_dereference_bh")) {
+    rewriter.replaceOpWithNewOp<RCUDereferenceBH>(exp, exp.getType(0),
+                                                  p_clone->getResult(0));
+  } else if (has_name(exp, "rcu_dereference_sched")) {
+    rewriter.replaceOpWithNewOp<RCUDereferenceSched>(exp, exp.getType(0),
+                                                     p_clone->getResult(0));
+  } else {
+    assert(!"Unexpected rcu_dereference() type in rewrite_rcu_dereference()");
+  }
+  return mlir::success();
+}
+
+mlir::LogicalResult
+rewrite_rcu_dereference_check(macroni::MacroExpansion exp,
+                              mlir::PatternRewriter &rewriter) {
+  if (!is_rcu_dereference_check(exp)) {
+    return mlir::failure();
+  }
+  mlir::Operation *p = nullptr;
+  mlir::Operation *c = nullptr;
+  exp.getExpansion().walk([&](mlir::Operation *op) {
+    if (auto mp = mlir::dyn_cast<macroni::MacroParameter>(op)) {
+      if (has_name(mp, "p")) {
+        p = op;
+      } else if (has_name(mp, "c")) {
+        c = op;
+      }
+    }
+  });
```

Comment:
Didn't I see something like this somewhere? Maybe built it in as method/set of standalone utilities.